### ISSUE TITLE
ICE 1.73 nightly

### ISSUE THREAD

**Issue by sftse** [rust-lang/rust#115051]
### Code

```Rust
trait Q<I>: Sized {
    type S;
}

trait P<I> {
    type X: Q<I>;
    fn a(_: &mut M<I, <Self::X as Q<I>>::S>)
    where
        Self: Sized,
    {
    }
}

struct R<A>(A);

impl<I, E: Q<I>, A> P<I> for R<A>
where
    A: P<I, X = E>,
{
    type X = E;

    fn a(_: &mut M<I, <Self::X as Q<I>>::S>)
    where
        Self: Sized,
    {
    }
}

struct M<I, S, Iter: Iterator<Item = S> + ?Sized = dyn Iterator<Item = S>>(I, S, Iter);

```

### Meta

Builds on
`rustc --version --verbose`:
```
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5
```

ICE on
```
rustc 1.73.0-nightly (28eb857b9 2023-08-12)
binary: rustc
commit-hash: 28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1
commit-date: 2023-08-12
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 17.0.0
```

### Error output

```
delayed span bug: unresolved inference variable in outlives: ?8t
   0: <rustc_errors::HandlerInner>::emit_diagnostic
   1: <rustc_session::session::Session>::delay_span_bug::<rustc_span::span_encoding::Span, alloc::string::String>
   2: <rustc_infer::infer::outlives::obligations::TypeOutlives<&rustc_infer::infer::InferCtxt>>::components_must_outlive
   3: <rustc_infer::infer::InferCtxt>::resolve_regions
   4: rustc_hir_analysis::check::compare_impl_item::compare_method_predicate_entailment
   5: rustc_hir_analysis::check::check::check_mod_item_types
   6: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::check_mod_item_types::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
   7: <rustc_query_impl::query_impl::check_mod_item_types::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId)>>::call_once
   8: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
   9: rustc_query_impl::query_impl::check_mod_item_types::get_query_non_incr::__rust_end_short_backtrace
  10: <rustc_middle::hir::map::Map>::for_each_module::<rustc_hir_analysis::check_crate::{closure#6}::{closure#0}>
  11: rustc_hir_analysis::check_crate
  12: rustc_interface::passes::analysis
  13: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 1]>>
  14: <rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
  15: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
  16: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  17: <rustc_middle::ty::context::GlobalCtxt>::enter::<rustc_driver_impl::run_compiler::{closure#1}::{closure#2}::{closure#6}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  18: rustc_span::set_source_map::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  19: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  20: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/alloc/src/boxed.rs:2007:9
  22: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/alloc/src/boxed.rs:2007:9
  23: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/std/src/sys/unix/thread.rs:108:17
  24: <unknown>
  25: <unknown>

delayed span bug: unresolved inference variable in outlives: ?7t
   0: <rustc_errors::HandlerInner>::emit_diagnostic
   1: <rustc_session::session::Session>::delay_span_bug::<rustc_span::span_encoding::Span, alloc::string::String>
   2: <rustc_infer::infer::outlives::obligations::TypeOutlives<&rustc_infer::infer::InferCtxt>>::components_must_outlive
   3: <rustc_infer::infer::InferCtxt>::resolve_regions
   4: rustc_hir_analysis::check::compare_impl_item::compare_method_predicate_entailment
   5: rustc_hir_analysis::check::check::check_mod_item_types
   6: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::check_mod_item_types::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
   7: <rustc_query_impl::query_impl::check_mod_item_types::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, rustc_span::def_id::LocalDefId)>>::call_once
   8: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
   9: rustc_query_impl::query_impl::check_mod_item_types::get_query_non_incr::__rust_end_short_backtrace
  10: <rustc_middle::hir::map::Map>::for_each_module::<rustc_hir_analysis::check_crate::{closure#6}::{closure#0}>
  11: rustc_hir_analysis::check_crate
  12: rustc_interface::passes::analysis
  13: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 1]>>
  14: <rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, ())>>::call_once
  15: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 1]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
  16: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  17: <rustc_middle::ty::context::GlobalCtxt>::enter::<rustc_driver_impl::run_compiler::{closure#1}::{closure#2}::{closure#6}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  18: rustc_span::set_source_map::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  19: std::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  20: <<std::thread::Builder>::spawn_unchecked_<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/alloc/src/boxed.rs:2007:9
  22: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/alloc/src/boxed.rs:2007:9
  23: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/28eb857b9504bd05bbed0cf8af8e825fbdbb1fa1/library/std/src/sys/unix/thread.rs:108:17
  24: <unknown>
  25: <unknown>



rustc version: 1.73.0-nightly (28eb857b9 2023-08-12)
platform: x86_64-unknown-linux-gnu

query stack during panic:
end of query stack
```

**Comment by sftse** [comment]
```rust
trait Q {
    type S;
}

trait P {
    type X: Q;
    fn a(_: &<Self::X as Q>::S)
    {
    }
}

struct R<A>(A);

impl<E: Q, A> P for R<A>
where
    A: P<X = E>,
{
    type X = E;

    fn a(_: &<Self::X as Q>::S)
    {
    }
}
```

**Comment by apiraino** [comment]
@sftse I'm afraid I can't reproduce the ICE with the provided MCVE. Can you post a bit more info on how to reproduce it? thanks

@darklyspaced did you remove the `I-prioritize` label intentionally? We use it to mark regressions that have not yet been estimated a priority. I'm going to reapply it.

@rustbot prioritize

**Comment by compiler-errors** [comment]
This is likely fixed by #115039.

**Comment by apiraino** [comment]
@sftse I've tried compiling the MCVE you provided [in this comment](https://github.com/rust-lang/rust/issues/115051#issuecomment-1686187353) by downloading the rustc version you indicated (1.73.0-nightly (28eb857b9 2023-08-12)) and then run:
```
$ rustc +nightly-2023-08-12-x86_64-unknown-linux-gnu --version
rustc 1.73.0-nightly (a6f8aa5a0 2023-08-11)

$ rustc +nightly-2023-08-12-x86_64-unknown-linux-gnu main.rs ; ./main
```

On my Linux workstation that does not emit any ICE so clearly I'm missing some context. What's the correct command line you run to compile that MCVE to make it trigger the ICE? Unsure why you mention the "chumsky" crate, at a cursory look I don't identify how this is relevant here. Or is it? thanks

**Comment by sftse** [comment]
The commit you listed, of your rustc, does not match the commit I listed. It seems quite strange that the command line you executed would respond with the rustc from a day earlier than you requested, is this a bug?

**Comment by apiraino** [comment]
yes, my bad. the bisection was slightly confusing because this has already been fixed in #115039 ([comment](https://github.com/rust-lang/rust/issues/115051#issuecomment-1686918448)), so latest nightly already don't reproduce this issue.

**Comment by JohnTitor** [comment]
Triage: Let's close this issue in favor of the PR mentioned above. If the regression test added by it doesn't cover this case, please re-open!
